File: examples/csharp/avalonia/NotesShowcase/ViewModels/NotebookVM.cs
using VMx.Builders;
using VMx.Capabilities;
using VMx.Components;
using VMx.Messages;
using VMx.Services;
using NotesShowcase.Models;

namespace NotesShowcase.ViewModels;

/// <summary>
/// Leaf VM for a notebook tree node.
///
/// Capabilities (plan §3.a.6, scenario §6.2):
///   <see cref="ISelectable"/>, <see cref="IExpandable"/>, <see cref="ICollapsible"/>,
///   <see cref="IExpansionTogglable"/>, <see cref="IReconstructable"/>.
///
/// Implemented as a direct subclass of <see cref="ComponentVMBase"/> rather than
/// <see cref="ComponentVM{M}"/> because the latter is sealed and we need to
/// layer the capability mix-ins. The model is held as a read-only field
/// (notebooks are renamed via repo + re-populate, not mutated in-place).
/// </summary>
public sealed class NotebookVM
    : ComponentVMBase,
      ISelectable,
      IExpandable,
      ICollapsible,
      IExpansionTogglable,
      IReconstructable
{
    private readonly ExpandableState _expansion;
    private NotebookModel _model;
    private Func<NotebookVM, IReadOnlyList<NotebookVM>>? _childrenGetter;

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>
    /// Public accessor for the underlying <see cref="IMessageHub"/> — exposed
    /// on the example VMs (not the library) so views and tests can subscribe
    /// without reflection. The view layer uses this to drive its
    /// <c>BindableVm</c> adapter (see Phase 4.a).
    /// </summary>
    public new IMessageHub Hub => base.Hub;

    /// <summary>Current notebook model.</summary>
    public NotebookModel Model
    {
        get => _model;
        set
        {
            if (EqualityComparer<NotebookModel>.Default.Equals(_model, value)) return;
            _model = value;
            Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Model)));
            RaisePropertyChanged(nameof(Model));
            RaisePropertyChanged(nameof(NotebookName));
            Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(NotebookName)));
        }
    }

    /// <summary>Notebook display name (derived from <see cref="Model"/>).</summary>
    public string NotebookName => _model.Name;

    /// <summary>
    /// Child notebook VMs (parent_id walk via the owner-supplied getter).
    /// Returns an empty list when no getter has been wired (standalone VMs do
    /// not know about siblings). Phase 5.a binding gap #2 (parity with
    /// Phase 5.b ``children_getter``): the Avalonia <c>TreeView</c>'s
    /// <c>TreeDataTemplate</c> binds <c>ItemsSource="{Binding Children}"</c>
    /// so nested notebooks (Specs under Work) appear under their parent.
    /// </summary>
    public IReadOnlyList<NotebookVM> Children
        => _childrenGetter is null ? Array.Empty<NotebookVM>() : _childrenGetter(this);

    /// <summary>
    /// Late-binds the children resolver (used by <see cref="NotebooksRootVM"/>
    /// after each <c>PopulateAsync</c> / <c>AddNotebookAsync</c>). Emits
    /// <see cref="PropertyChangedMessage{TSender}"/> for <c>Children</c> so
    /// already-bound tree views refresh.
    /// </summary>
    public void SetChildrenGetter(Func<NotebookVM, IReadOnlyList<NotebookVM>>? getter)
    {
        _childrenGetter = getter;
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Children)));
        RaisePropertyChanged(nameof(Children));
    }

    /// <summary>
    /// Re-emits a <c>Children</c> change notification (called by
    /// <see cref="NotebooksRootVM"/> whenever the flat collection mutates so
    /// already-bound parents refresh their child list).
    /// </summary>
    internal void NotifyChildrenChanged()
    {
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Children)));
        RaisePropertyChanged(nameof(Children));
    }

    // ── ISelectable / IExpandable / ICollapsible / IExpansionTogglable ──────

    /// <inheritdoc cref="IExpandable.IsExpanded"/>
    public bool IsExpanded => _expansion.IsExpanded;

    /// <inheritdoc cref="IExpandable.CanExpand"/>
    public bool CanExpand() => _expansion.CanExpand();

    /// <inheritdoc cref="IExpandable.Expand"/>
    public void Expand()
    {
        if (!_expansion.CanExpand()) return;
        _expansion.Expand();
        EmitExpansionChange();
    }

    /// <inheritdoc cref="ICollapsible.CanCollapse"/>
    public bool CanCollapse() => _expansion.CanCollapse();

    /// <inheritdoc cref="ICollapsible.Collapse"/>
    public void Collapse()
    {
        if (!_expansion.CanCollapse()) return;
        _expansion.Collapse();
        EmitExpansionChange();
    }

    /// <inheritdoc cref="IExpansionTogglable.CanToggleExpansion"/>
    public bool CanToggleExpansion() => _expansion.CanToggleExpansion();

    /// <inheritdoc cref="IExpansionTogglable.ToggleExpansion"/>
    public void ToggleExpansion()
    {
        if (_expansion.IsExpanded) Collapse(); else Expand();
    }

    private void EmitExpansionChange()
    {
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(IsExpanded)));
        RaisePropertyChanged(nameof(IsExpanded));
    }

    // Selection delegates to base ComponentVMBase Select/Deselect plumbing.

    private NotebookVM(
        string name,
        string hint,
        NotebookModel model,
        IMessageHub hub,
        IDispatcher dispatcher,
        bool initiallyExpanded,
        Func<NotebookVM, IReadOnlyList<NotebookVM>>? childrenGetter)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _model = model;
        _expansion = new ExpandableState(initiallyExpanded);
        _childrenGetter = childrenGetter;
    }

    /// <inheritdoc/>
    protected override void OnDispose()
    {
        _expansion.Dispose();
        base.OnDispose();
    }

    /// <summary>Returns a new empty builder for <see cref="NotebookVM"/>.</summary>
    public static NotebookVMBuilder Builder() => NotebookVMBuilder.Empty;

    /// <summary>Immutable fluent builder for <see cref="NotebookVM"/> (spec ch. 10).</summary>
    public sealed class NotebookVMBuilder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly NotebookModel? _model;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly bool _initiallyExpanded;
        private readonly Func<NotebookVM, IReadOnlyList<NotebookVM>>? _childrenGetter;

        internal static readonly NotebookVMBuilder Empty = new();

        private NotebookVMBuilder() { _hint = ""; }

        private NotebookVMBuilder(
            string? name,
            string hint,
            NotebookModel? model,
            IMessageHub? hub,
            IDispatcher? dispatcher,
            bool initiallyExpanded,
            Func<NotebookVM, IReadOnlyList<NotebookVM>>? childrenGetter)
        {
            _name = name;
            _hint = hint;
            _model = model;
            _hub = hub;
            _dispatcher = dispatcher;
            _initiallyExpanded = initiallyExpanded;
            _childrenGetter = childrenGetter;
        }

        /// <summary>Sets the required VM Name.</summary>
        public NotebookVMBuilder Name(string name)
            => new(name, _hint, _model, _hub, _dispatcher, _initiallyExpanded, _childrenGetter);

        /// <summary>Sets the optional Hint.</summary>
        public NotebookVMBuilder Hint(string hint)
            => new(_name, hint, _model, _hub, _dispatcher, _initiallyExpanded, _childrenGetter);

        /// <summary>Sets the required notebook model.</summary>
        public NotebookVMBuilder Model(NotebookModel model)
            => new(_name, _hint, model, _hub, _dispatcher, _initiallyExpanded, _childrenGetter);

        /// <summary>Sets the required Services (hub + dispatcher).</summary>
        public NotebookVMBuilder Services(IMessageHub hub, IDispatcher dispatcher)
            => new(_name, _hint, _model, hub, dispatcher, _initiallyExpanded, _childrenGetter);

        /// <summary>Sets the optional initial expansion state (default false).</summary>
        public NotebookVMBuilder InitiallyExpanded(bool initiallyExpanded)
            => new(_name, _hint, _model, _hub, _dispatcher, initiallyExpanded, _childrenGetter);

        /// <summary>
        /// Sets the optional children-getter callback. Phase 5.a binding gap #2:
        /// the owning <see cref="NotebooksRootVM"/> wires this so each notebook
        /// can resolve its children from the flat collection.
        /// </summary>
        public NotebookVMBuilder ChildrenGetter(Func<NotebookVM, IReadOnlyList<NotebookVM>> getter)
            => new(_name, _hint, _model, _hub, _dispatcher, _initiallyExpanded, getter);

        /// <summary>Builds the VM after validating required fields.</summary>
        public NotebookVM Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_model, "Model");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            return new NotebookVM(_name!, _hint, _model!, _hub!, _dispatcher!, _initiallyExpanded, _childrenGetter);
        }
    }
}

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/ViewModels/NotebookVM.swift
//
// NotebookVM — leaf view-model for a notebook tree node.
//
// Capabilities (scenario §6.2):
//   `Selectable`, `Deselectable`, `Expandable`, `Collapsible`,
//   `ExpansionTogglable`, `Reconstructable`.
//
// Implemented as a direct `ComponentVMBase` subclass (not `ComponentVM<M>`)
// so the capability mix-ins can be layered without the sealed generic wrapper.
// Cross-module subclassing enabled by ADR-0066: `hub`, `dispatcher`, and
// `_raisePropertyChanged` are `public` on the base class.
//
import Foundation
import Combine
import VMx

/// Leaf view-model for a notebook tree node.
///
/// Mirrors the C# Avalonia `NotebookVM`. Holds a `NotebookModel` (equality-
/// guarded setter), an owner-supplied `childrenGetter`, and delegates
/// expansion state to the library's `ExpandableState` helper.
public final class NotebookVM: ComponentVMBase,
                                Selectable, Deselectable,
                                Expandable, Collapsible, ExpansionTogglable {
    // `Reconstructable` is satisfied automatically via
    // `extension ComponentVMBase: Constructable, Destructable, Reconstructable {}`.

    // ── Private state ──────────────────────────────────────────────────────

    private var _model: NotebookModel
    private let _expansion: ExpandableState
    private var _childrenGetter: ((NotebookVM) -> [NotebookVM])?

    // ── Public properties ──────────────────────────────────────────────────

    /// Current notebook model. Equality-guarded; emits `PropertyChangedMessage`
    /// for `"model"` and `"notebookName"` on change.
    public var model: NotebookModel {
        get { _model }
        set {
            guard _model != newValue else { return }
            _model = newValue
            hub.send(PropertyChangedMessage(sender: self, senderName: name, propertyName: "model"))
            _raisePropertyChanged("model")
            hub.send(PropertyChangedMessage(sender: self, senderName: name, propertyName: "notebookName"))
            _raisePropertyChanged("notebookName")
        }
    }

    /// Notebook display name (derived from `model`).
    public var notebookName: String { _model.name }

    /// Child notebook VMs resolved via the owner-supplied getter.
    /// Returns an empty array when no getter has been wired (standalone VMs).
    public var children: [NotebookVM] {
        _childrenGetter.map { $0(self) } ?? []
    }

    // ── Children resolver ──────────────────────────────────────────────────

    /// Late-binds the children resolver (called by `NotebooksRootVM` after
    /// each `populateAsync` / `addNotebookAsync`). Emits a `PropertyChangedMessage`
    /// for `"children"` so already-bound views refresh.
    public func setChildrenGetter(_ getter: ((NotebookVM) -> [NotebookVM])?) {
        _childrenGetter = getter
        hub.send(PropertyChangedMessage(sender: self, senderName: name, propertyName: "children"))
        _raisePropertyChanged("children")
    }

    /// Re-emits a `"children"` change notification (called by `NotebooksRootVM`
    /// whenever the flat collection mutates so already-bound parents refresh).
    public func notifyChildrenChanged() {
        hub.send(PropertyChangedMessage(sender: self, senderName: name, propertyName: "children"))
        _raisePropertyChanged("children")
    }

    // ── Selectable / Deselectable (delegated to ComponentVMBase) ──────────
    // ComponentVMBase already provides canSelect/select/canDeselect/deselect;
    // the protocol conformance declarations above satisfy the contracts.

    // ── Expandable ─────────────────────────────────────────────────────────

    public var isExpanded: Bool { _expansion.isExpanded }

    public func canExpand() -> Bool { _expansion.canExpand() }

    public func expand() {
        guard _expansion.canExpand() else { return }
        _expansion.expand()
        _emitExpansionChange()
    }

    // ── Collapsible ────────────────────────────────────────────────────────

    public func canCollapse() -> Bool { _expansion.canCollapse() }

    public func collapse() {
        guard _expansion.canCollapse() else { return }
        _expansion.collapse()
        _emitExpansionChange()
    }

    // ── ExpansionTogglable ─────────────────────────────────────────────────

    public func canToggleExpansion() -> Bool { _expansion.canToggleExpansion() }

    public func toggleExpansion() {
        if _expansion.isExpanded { collapse() } else { expand() }
    }

    private func _emitExpansionChange() {
        hub.send(PropertyChangedMessage(sender: self, senderName: name, propertyName: "isExpanded"))
        _raisePropertyChanged("isExpanded")
    }

    // ── Init ───────────────────────────────────────────────────────────────

    private init(
        name: String,
        hint: String,
        model: NotebookModel,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        initiallyExpanded: Bool,
        childrenGetter: ((NotebookVM) -> [NotebookVM])?
    ) {
        _model = model
        _expansion = ExpandableState(initiallyExpanded: initiallyExpanded)
        _childrenGetter = childrenGetter
        super.init(name: name, hint: hint, hub: hub, dispatcher: dispatcher)
    }

    // ── Dispose ────────────────────────────────────────────────────────────

    override public func _onDispose() {
        _expansion.dispose()
        super._onDispose()
    }

    // ── Builder ────────────────────────────────────────────────────────────

    /// Returns a new empty builder for `NotebookVM`.
    public static func builder() -> NotebookVMBuilder {
        NotebookVMBuilder()
    }

    /// Immutable fluent builder for `NotebookVM` (spec ch. 10).
    ///
    /// Required fields: `name(_:)`, `model(_:)`, `services(hub:dispatcher:)`.
    /// Optional fields: `initiallyExpanded(_:)` (default `false`), `childrenGetter(_:)`.
    public struct NotebookVMBuilder {
        private var _name: String?
        private var _hint: String = ""
        private var _model: NotebookModel?
        private var _hub: MessageHubProtocol?
        private var _dispatcher: Dispatcher?
        private var _initiallyExpanded: Bool = false
        private var _childrenGetter: ((NotebookVM) -> [NotebookVM])?

        fileprivate init() {}

        /// Sets the required VM name.
        public func name(_ value: String) -> NotebookVMBuilder {
            var copy = self; copy._name = value; return copy
        }

        /// Sets the optional hint.
        public func hint(_ value: String) -> NotebookVMBuilder {
            var copy = self; copy._hint = value; return copy
        }

        /// Sets the required notebook model.
        public func model(_ value: NotebookModel) -> NotebookVMBuilder {
            var copy = self; copy._model = value; return copy
        }

        /// Sets the required services (hub + dispatcher).
        public func services(hub: MessageHubProtocol, dispatcher: Dispatcher) -> NotebookVMBuilder {
            var copy = self
            copy._hub = hub
            copy._dispatcher = dispatcher
            return copy
        }

        /// Sets the optional initial expansion state (default `false`).
        public func initiallyExpanded(_ value: Bool = true) -> NotebookVMBuilder {
            var copy = self; copy._initiallyExpanded = value; return copy
        }

        /// Sets the optional children-getter callback. The owning
        /// `NotebooksRootVM` wires this so each notebook can resolve its
        /// children from the flat collection.
        public func childrenGetter(_ getter: @escaping (NotebookVM) -> [NotebookVM]) -> NotebookVMBuilder {
            var copy = self; copy._childrenGetter = getter; return copy
        }

        /// Validates required fields and constructs a `NotebookVM`.
        ///
        /// - Throws: `BuilderValidationError` if `name`, `model`, or `services` are missing.
        public func build() throws -> NotebookVM {
            guard let name = _name else {
                throw BuilderValidationError(missingField: "name")
            }
            guard let model = _model else {
                throw BuilderValidationError(missingField: "model")
            }
            guard let hub = _hub else {
                throw BuilderValidationError(missingField: "hub")
            }
            guard let dispatcher = _dispatcher else {
                throw BuilderValidationError(missingField: "dispatcher")
            }
            return NotebookVM(
                name: name,
                hint: _hint,
                model: model,
                hub: hub,
                dispatcher: dispatcher,
                initiallyExpanded: _initiallyExpanded,
                childrenGetter: _childrenGetter
            )
        }
    }
}
